parse/lex.go

```diff
@@ -32,18 +36,22 @@ func New(input string) *Lexer {
 
 func (l *Lexer) run(input string) {
 
-	state := newState(input)
-	var token Token
+	state := state{
+		input: input,
+	}
 
 	for stateFn := DefaultStateFn; stateFn != nil; {
-		stateFn, state, token = stateFn(state)
-		l.tokench <- token
+		stateFn = stateFn(state, l)
```

Comment:
I think stateFn should have address of `state` if our lexer is going to have `state` as local variable